File: coding-interview/python/p06_graph.py
"""面试手撕代码（06）：图 —— BFS / DFS / 拓扑排序 / 并查集 / 最短路 —— 配套解法与测试。

https://arganzheng.life/coding-interview-graph-bfs-dfs-topological-union-find.html
"""
from __future__ import annotations

import heapq
import unittest
from collections import defaultdict, deque
from typing import Dict, List, Optional

DIRS = ((1, 0), (-1, 0), (0, 1), (0, -1))


def num_islands(grid: List[List[str]]) -> int:
    """LC 200. 网格 DFS，访问过的格子直接改成 '0'（沉岛）。O(mn) / O(mn) 栈。"""
    m, n = len(grid), len(grid[0])

    def sink(i, j):
        if not (0 <= i < m and 0 <= j < n) or grid[i][j] != "1":
            return
        grid[i][j] = "0"
        for di, dj in DIRS:
            sink(i + di, j + dj)

    count = 0
    for i in range(m):
        for j in range(n):
            if grid[i][j] == "1":
                count += 1
                sink(i, j)
    return count


def max_area_of_island(grid: List[List[int]]) -> int:
    """LC 695. 同一模板，DFS 返回面积。"""
    m, n = len(grid), len(grid[0])

    def area(i, j):
        if not (0 <= i < m and 0 <= j < n) or grid[i][j] != 1:
            return 0
        grid[i][j] = 0
        return 1 + sum(area(i + di, j + dj) for di, dj in DIRS)

    return max((area(i, j) for i in range(m) for j in range(n)), default=0)


def oranges_rotting(grid: List[List[int]]) -> int:
    """LC 994. 多源 BFS：所有腐烂橘子同时入队，按层计分钟。O(mn) / O(mn)。"""
    m, n = len(grid), len(grid[0])
    q = deque()
    fresh = 0
    for i in range(m):
        for j in range(n):
            if grid[i][j] == 2:
                q.append((i, j))
            elif grid[i][j] == 1:
                fresh += 1
    minutes = 0
    while q and fresh:
        for _ in range(len(q)):
            i, j = q.popleft()
            for di, dj in DIRS:
                x, y = i + di, j + dj
                if 0 <= x < m and 0 <= y < n and grid[x][y] == 1:
                    grid[x][y] = 2
                    fresh -= 1
                    q.append((x, y))
        minutes += 1
    return -1 if fresh else minutes


def can_finish(num_courses: int, prerequisites: List[List[int]]) -> bool:
    """LC 207. Kahn 拓扑：入度为 0 的进队，出队时给后继减入度。O(V+E)。"""
    return len(find_order(num_courses, prerequisites)) == num_courses


def find_order(num_courses: int, prerequisites: List[List[int]]) -> List[int]:
    """LC 210. 同上，返回顺序；有环则返回 []。"""
    graph = defaultdict(list)
    indeg = [0] * num_courses
    for a, b in prerequisites:              # b -> a
        graph[b].append(a)
        indeg[a] += 1
    q = deque(i for i in range(num_courses) if indeg[i] == 0)
    order = []
    while q:
        u = q.popleft()
        order.append(u)
        for v in graph[u]:
            indeg[v] -= 1
            if indeg[v] == 0:
                q.append(v)
    return order if len(order) == num_courses else []


def can_finish_dfs(num_courses: int, prerequisites: List[List[int]]) -> bool:
    """LC 207 的 DFS 三色标记版：灰色再遇到即有环。"""
    graph = defaultdict(list)
    for a, b in prerequisites:
        graph[b].append(a)
    color = [0] * num_courses               # 0 白 1 灰 2 黑

    def dfs(u):
        color[u] = 1
        for v in graph[u]:
            if color[v] == 1 or (color[v] == 0 and not dfs(v)):
                return False
        color[u] = 2
        return True

    return all(color[i] or dfs(i) for i in range(num_courses))


class GraphNode:
    def __init__(self, val: int = 0, neighbors: Optional[List["GraphNode"]] = None):
        self.val, self.neighbors = val, neighbors or []


def clone_graph(node: Optional[GraphNode]) -> Optional[GraphNode]:
    """LC 133. 哈希 old -> new 兼作 visited，DFS。O(V+E)。"""
    if not node:
        return None
    mapping: Dict[GraphNode, GraphNode] = {}

    def dfs(u):
        if u in mapping:
            return mapping[u]
        copy = GraphNode(u.val)
        mapping[u] = copy                    # 先登记再递归，处理环
        copy.neighbors = [dfs(v) for v in u.neighbors]
        return copy

    return dfs(node)


def ladder_length(begin: str, end: str, word_list: List[str]) -> int:
    """LC 127. 双向 BFS：每次扩展较小的一侧。O(N * L * 26)。"""
    words = set(word_list)
    if end not in words:
        return 0
    front, back = {begin}, {end}
    steps = 1
    while front and back:
        if len(front) > len(back):
            front, back = back, front
        nxt = set()
        for w in front:
            for i in range(len(w)):
                for c in "abcdefghijklmnopqrstuvwxyz":
                    cand = w[:i] + c + w[i + 1:]
                    if cand in back:
                        return steps + 1
                    if cand in words:
                        words.remove(cand)   # 用过就删，兼作 visited
                        nxt.add(cand)
        front = nxt
        steps += 1
    return 0


class UnionFind:
    """路径压缩 + 按大小合并；find 均摊近 O(1)。"""

    def __init__(self, n: int):
        self.parent = list(range(n))
        self.size = [1] * n
        self.count = n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]   # 路径减半
            x = self.parent[x]
        return x

    def union(self, a: int, b: int) -> bool:
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        if self.size[ra] < self.size[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        self.size[ra] += self.size[rb]
        self.count -= 1
        return True


def find_circle_num(is_connected: List[List[int]]) -> int:
    """LC 547. 并查集数连通块。O(n^2 α)。"""
    n = len(is_connected)
    uf = UnionFind(n)
    for i in range(n):
        for j in range(i + 1, n):
            if is_connected[i][j]:
                uf.union(i, j)
    return uf.count


def accounts_merge(accounts: List[List[str]]) -> List[List[str]]:
    """LC 721. 邮箱映射到账户下标，同账户邮箱合并；输出按根聚合并排序。"""
    uf = UnionFind(len(accounts))
    owner: Dict[str, int] = {}
    for i, acc in enumerate(accounts):
        for email in acc[1:]:
            if email in owner:
                uf.union(i, owner[email])
            else:
                owner[email] = i
    groups = defaultdict(list)
    for email, i in owner.items():
        groups[uf.find(i)].append(email)
    return [[accounts[r][0]] + sorted(emails) for r, emails in groups.items()]


def network_delay_time(times: List[List[int]], n: int, k: int) -> int:
    """LC 743. Dijkstra（堆 + 懒删除）。O(E log E)。"""
    graph = defaultdict(list)
    for u, v, w in times:
        graph[u].append((v, w))
    dist = {k: 0}
    heap = [(0, k)]
    while heap:
        d, u = heapq.heappop(heap)
        if d > dist.get(u, float("inf")):
            continue                         # 过期条目
        for v, w in graph[u]:
            nd = d + w
            if nd < dist.get(v, float("inf")):
                dist[v] = nd
                heapq.heappush(heap, (nd, v))
    return max(dist.values()) if len(dist) == n else -1


def shortest_path_binary_matrix(grid: List[List[int]]) -> int:
    """LC 1091. 八方向网格 BFS。"""
    n = len(grid)
    if grid[0][0] or grid[-1][-1]:
        return -1
    q = deque([(0, 0, 1)])
    grid[0][0] = 1
    while q:
        i, j, d = q.popleft()
        if i == j == n - 1:
            return d
        for di in (-1, 0, 1):
            for dj in (-1, 0, 1):
                x, y = i + di, j + dj
                if 0 <= x < n and 0 <= y < n and grid[x][y] == 0:
                    grid[x][y] = 1
                    q.append((x, y, d + 1))
    return -1


class Tests(unittest.TestCase):
    def test_islands(self):
        g = [list("11000"), list("11000"), list("00100"), list("00011")]
        self.assertEqual(num_islands(g), 3)
        self.assertEqual(max_area_of_island([[0, 1, 1, 0], [0, 1, 0, 0], [1, 0, 0, 1]]), 3)

    def test_oranges(self):
        self.assertEqual(oranges_rotting([[2, 1, 1], [1, 1, 0], [0, 1, 1]]), 4)
        self.assertEqual(oranges_rotting([[2, 1, 1], [0, 1, 1], [1, 0, 1]]), -1)
        self.assertEqual(oranges_rotting([[0, 2]]), 0)

    def test_topo(self):
        self.assertTrue(can_finish(2, [[1, 0]]))
        self.assertFalse(can_finish(2, [[1, 0], [0, 1]]))
        self.assertEqual(find_order(4, [[1, 0], [2, 0], [3, 1], [3, 2]]), [0, 1, 2, 3])
        self.assertTrue(can_finish_dfs(3, [[1, 0], [2, 1]]))
        self.assertFalse(can_finish_dfs(3, [[1, 0], [2, 1], [0, 2]]))

    def test_clone(self):
        a, b, c = GraphNode(1), GraphNode(2), GraphNode(3)
        a.neighbors, b.neighbors, c.neighbors = [b, c], [a, c], [a, b]
        a2 = clone_graph(a)
        self.assertIsNot(a2, a)
        self.assertEqual(sorted(x.val for x in a2.neighbors), [2, 3])
        self.assertIs(a2.neighbors[0].neighbors[0], a2)

    def test_ladder(self):
        self.assertEqual(ladder_length("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]), 5)
        self.assertEqual(ladder_length("hit", "cog", ["hot", "dot", "dog", "lot", "log"]), 0)

    def test_union_find(self):
        self.assertEqual(find_circle_num([[1, 1, 0], [1, 1, 0], [0, 0, 1]]), 2)
        out = accounts_merge([["John", "johnsmith@mail.com", "john_newyork@mail.com"],
                              ["John", "johnsmith@mail.com", "john00@mail.com"],
                              ["Mary", "mary@mail.com"], ["John", "johnnybravo@mail.com"]])
        self.assertEqual(sorted(out), sorted([["John", "john00@mail.com", "john_newyork@mail.com", "johnsmith@mail.com"],
                                               ["Mary", "mary@mail.com"], ["John", "johnnybravo@mail.com"]]))

    def test_dijkstra(self):
        self.assertEqual(network_delay_time([[2, 1, 1], [2, 3, 1], [3, 4, 1]], 4, 2), 2)
        self.assertEqual(network_delay_time([[1, 2, 1]], 2, 2), -1)

    def test_binary_matrix(self):
        self.assertEqual(shortest_path_binary_matrix([[0, 0, 0], [1, 1, 0], [1, 1, 0]]), 4)
        self.assertEqual(shortest_path_binary_matrix([[1, 0, 0], [1, 1, 0], [1, 1, 0]]), -1)


if __name__ == "__main__":
    unittest.main(verbosity=1)

File: coding-interview/java/P06Graph.java
import java.util.*;

/** 面试手撕代码（06）：图 —— BFS / DFS / 拓扑 / 并查集 / 最短路。运行：java -ea P06Graph */
public class P06Graph {

    static final int[][] DIRS = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    /** LC 200. 沉岛 DFS。 */
    static int numIslands(char[][] g) {
        int count = 0;
        for (int i = 0; i < g.length; i++)
            for (int j = 0; j < g[0].length; j++)
                if (g[i][j] == '1') {
                    count++;
                    sink(g, i, j);
                }
        return count;
    }

    private static void sink(char[][] g, int i, int j) {
        if (i < 0 || j < 0 || i >= g.length || j >= g[0].length || g[i][j] != '1') return;
        g[i][j] = '0';
        for (int[] d : DIRS) sink(g, i + d[0], j + d[1]);
    }

    /** LC 994. 多源 BFS 按层计时。 */
    static int orangesRotting(int[][] g) {
        int m = g.length, n = g[0].length, fresh = 0, minutes = 0;
        Deque<int[]> q = new ArrayDeque<>();
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++) {
                if (g[i][j] == 2) q.add(new int[] {i, j});
                else if (g[i][j] == 1) fresh++;
            }
        while (!q.isEmpty() && fresh > 0) {
            for (int s = q.size(); s > 0; s--) {
                int[] c = q.poll();
                for (int[] d : DIRS) {
                    int x = c[0] + d[0], y = c[1] + d[1];
                    if (x >= 0 && y >= 0 && x < m && y < n && g[x][y] == 1) {
                        g[x][y] = 2;
                        fresh--;
                        q.add(new int[] {x, y});
                    }
                }
            }
            minutes++;
        }
        return fresh == 0 ? minutes : -1;
    }

    /** LC 210（含 207）. Kahn 拓扑排序，有环返回空数组。 */
    static int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) graph.add(new ArrayList<>());
        int[] indeg = new int[numCourses];
        for (int[] p : prerequisites) {
            graph.get(p[1]).add(p[0]);
            indeg[p[0]]++;
        }
        Deque<Integer> q = new ArrayDeque<>();
        for (int i = 0; i < numCourses; i++) if (indeg[i] == 0) q.add(i);
        int[] order = new int[numCourses];
        int k = 0;
        while (!q.isEmpty()) {
            int u = q.poll();
            order[k++] = u;
            for (int v : graph.get(u)) if (--indeg[v] == 0) q.add(v);
        }
        return k == numCourses ? order : new int[0];
    }

    static boolean canFinish(int n, int[][] pre) {
        return findOrder(n, pre).length == n;
    }

    static class GraphNode {
        int val;
        List<GraphNode> neighbors = new ArrayList<>();

        GraphNode(int v) {
            val = v;
        }
    }

    /** LC 133. 哈希兼作 visited 的 DFS。 */
    static GraphNode cloneGraph(GraphNode node) {
        return node == null ? null : clone(node, new HashMap<>());
    }

    private static GraphNode clone(GraphNode u, Map<GraphNode, GraphNode> map) {
        if (map.containsKey(u)) return map.get(u);
        GraphNode copy = new GraphNode(u.val);
        map.put(u, copy);
        for (GraphNode v : u.neighbors) copy.neighbors.add(clone(v, map));
        return copy;
    }

    /** LC 127. 双向 BFS。 */
    static int ladderLength(String begin, String end, List<String> wordList) {
        Set<String> words = new HashSet<>(wordList);
        if (!words.contains(end)) return 0;
        Set<String> front = new HashSet<>(List.of(begin)), back = new HashSet<>(List.of(end));
        int steps = 1;
        while (!front.isEmpty() && !back.isEmpty()) {
            if (front.size() > back.size()) {
                Set<String> t = front;
                front = back;
                back = t;
            }
            Set<String> next = new HashSet<>();
            for (String w : front) {
                char[] cs = w.toCharArray();
                for (int i = 0; i < cs.length; i++) {
                    char orig = cs[i];
                    for (char c = 'a'; c <= 'z'; c++) {
                        cs[i] = c;
                        String cand = new String(cs);
                        if (back.contains(cand)) return steps + 1;
                        if (words.remove(cand)) next.add(cand);
                    }
                    cs[i] = orig;
                }
            }
            front = next;
            steps++;
        }
        return 0;
    }

    /** 并查集：路径压缩 + 按大小合并。 */
    static class UnionFind {
        int[] parent, size;
        int count;

        UnionFind(int n) {
            parent = new int[n];
            size = new int[n];
            count = n;
            for (int i = 0; i < n; i++) {
                parent[i] = i;
                size[i] = 1;
            }
        }

        int find(int x) {
            while (parent[x] != x) {
                parent[x] = parent[parent[x]];
                x = parent[x];
            }
            return x;
        }

        boolean union(int a, int b) {
            int ra = find(a), rb = find(b);
            if (ra == rb) return false;
            if (size[ra] < size[rb]) {
                int t = ra;
                ra = rb;
                rb = t;
            }
            parent[rb] = ra;
            size[ra] += size[rb];
            count--;
            return true;
        }
    }

    /** LC 547. */
    static int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        UnionFind uf = new UnionFind(n);
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++) if (isConnected[i][j] == 1) uf.union(i, j);
        return uf.count;
    }

    /** LC 721. */
    static List<List<String>> accountsMerge(List<List<String>> accounts) {
        UnionFind uf = new UnionFind(accounts.size());
        Map<String, Integer> owner = new HashMap<>();
        for (int i = 0; i < accounts.size(); i++)
            for (String email : accounts.get(i).subList(1, accounts.get(i).size())) {
                Integer j = owner.putIfAbsent(email, i);
                if (j != null) uf.union(i, j);
            }
        Map<Integer, TreeSet<String>> groups = new HashMap<>();
        for (Map.Entry<String, Integer> e : owner.entrySet())
            groups.computeIfAbsent(uf.find(e.getValue()), k -> new TreeSet<>()).add(e.getKey());
        List<List<String>> out = new ArrayList<>();
        for (Map.Entry<Integer, TreeSet<String>> e : groups.entrySet()) {
            List<String> acc = new ArrayList<>();
            acc.add(accounts.get(e.getKey()).get(0));
            acc.addAll(e.getValue());
            out.add(acc);
        }
        return out;
    }

    /** LC 743. Dijkstra（堆 + 懒删除）。 */
    static int networkDelayTime(int[][] times, int n, int k) {
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) graph.add(new ArrayList<>());
        for (int[] t : times) graph.get(t[0]).add(new int[] {t[1], t[2]});
        int[] dist = new int[n + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[k] = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        pq.offer(new int[] {0, k});
        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int d = cur[0], u = cur[1];
            if (d > dist[u]) continue;
            for (int[] e : graph.get(u)) {
                int nd = d + e[1];
                if (nd < dist[e[0]]) {
                    dist[e[0]] = nd;
                    pq.offer(new int[] {nd, e[0]});
                }
            }
        }
        int best = 0;
        for (int i = 1; i <= n; i++) {
            if (dist[i] == Integer.MAX_VALUE) return -1;
            best = Math.max(best, dist[i]);
        }
        return best;
    }

    public static void main(String[] args) {
        char[][] g = {
            "11000".toCharArray(),
            "11000".toCharArray(),
            "00100".toCharArray(),
            "00011".toCharArray()
        };
        assert numIslands(g) == 3;
        assert orangesRotting(new int[][] {{2, 1, 1}, {1, 1, 0}, {0, 1, 1}}) == 4;
        assert orangesRotting(new int[][] {{2, 1, 1}, {0, 1, 1}, {1, 0, 1}}) == -1;
        assert canFinish(2, new int[][] {{1, 0}}) && !canFinish(2, new int[][] {{1, 0}, {0, 1}});
        assert Arrays.equals(
                findOrder(4, new int[][] {{1, 0}, {2, 0}, {3, 1}, {3, 2}}), new int[] {0, 1, 2, 3});
        GraphNode a = new GraphNode(1), b = new GraphNode(2);
        a.neighbors.add(b);
        b.neighbors.add(a);
        GraphNode a2 = cloneGraph(a);
        assert a2 != a && a2.neighbors.get(0).neighbors.get(0) == a2;
        assert ladderLength("hit", "cog", List.of("hot", "dot", "dog", "lot", "log", "cog")) == 5;
        assert findCircleNum(new int[][] {{1, 1, 0}, {1, 1, 0}, {0, 0, 1}}) == 2;
        List<List<String>> merged =
                accountsMerge(
                        List.of(
                                List.of("John", "johnsmith@mail.com", "john_newyork@mail.com"),
                                List.of("John", "johnsmith@mail.com", "john00@mail.com"),
                                List.of("Mary", "mary@mail.com"),
                                List.of("John", "johnnybravo@mail.com")));
        assert merged.size() == 3;
        assert networkDelayTime(new int[][] {{2, 1, 1}, {2, 3, 1}, {3, 4, 1}}, 4, 2) == 2;
        assert networkDelayTime(new int[][] {{1, 2, 1}}, 2, 2) == -1;
        System.out.println("P06Graph OK");
    }
}
